<a href="https://colab.research.google.com/github/paascorb/MergeSortPython/blob/main/TrabajoEloy.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Trabajo Mergesort Python

En este apartado vamos a realizar la parte obligatoria del trabajo, donde vamos a programar el método de ordenación Mergesort en Python dse manera secuencial, paralela y usando Numba.

## Utilidades

Primero, antes de ponernos a programar los algoritmos, necesitamos programar los métodos auxiliares.

In [1]:
from random import *

def generar_datos(tamaño):
  '''Genera 'tamaño' datos aleatorios dcon valores entre -1000 y 1000
     Siempre genera la misma secuencia de valores'''

  seed(22) # Fijamos la semilla
  return [randint(-1000, 1000) for i in range(0, tamaño)]

Probamos la primera función.

In [2]:
generar_datos(3)

[962, 858, -713]

Seguimos con las siguientes.

In [3]:
def esta_ordenado_array(vector, tamaño, ti, tf):
  '''Comprueba si un array está ordenado
     Devolverá la primera posición que no está ordenada y -1 si el vector está ordenado'''

  aux = -1

  for i in range(0, tamaño-1):
    if(vector[i]>vector[i+1]):
      aux = i
      i = tamaño

  if(aux == -1):

    print("El array está ordenado\n\n")
    print("Tiempo invertido en ordenar: "+str(tf-ti)+" milisegundos \n\n")

  else:
    print("\nEl vector falla en ****  "+ str(aux) +"  ******")

  return aux

In [4]:
esta_ordenado_array([1,2,3,4], 4, 100, 102)

El array está ordenado


Tiempo invertido en ordenar: 2 milisegundos 




-1

In [5]:
esta_ordenado_array([1,5,3,4], 4, 100, 102)


El vector falla en ****  1  ******


1

In [6]:
def burbuja(v, izq, der):
  '''Ordenación secuencial por el método de la burbuja
     Se pasan los índices izq y ordena desde v[iz] hasta v[der-1]'''
  aux = 0
  for i in range(0, der-izq-1):
    for j in range(izq, der-i-1):
      if v[j] > v[j+1]:
        aux = v[j]
        v[j]=v[j+1]
        v[j+1]=aux

In [7]:
prueba = [3,2,1]
burbuja(prueba,0,3)
prueba

[1, 2, 3]

In [8]:
import time

def milisegundos():
  # Devuelve el tiempo en milisengundos desde la época Unix (01/01/1970)

	return round(time.time() * 1000)

In [9]:
milisegundos()

1686694868624

In [10]:
def mezclar(v , izq, med, der):
  '''Mezcla ordenada de dos trozos de un vector desde sus índices izq hasta der
     Primer trozo desde izq hasta med-1
     Segundo trozo desde med hasta der-1
     Cada trozo tiene que estar previamente ordenado'''

  aux= [0] * (der-izq)
  i=izq
  j=med
  k=0

  while i<med and j<der:
    if v[i] <= v[j]:
      aux[k] = v[i]
      i += 1
    else:
      aux[k] = v[j];
      j += 1
    k += 1

  while i<med:
    aux[k] = v[i]
    k += 1
    i += 1

  while j<der:
    aux[k] = v[j];
    k += 1
    j += 1

  for i in range(0, der-izq):
      v[i+izq] = aux[i]

## Python Secuencial

En esta sección vamos a programar el algoritmo Mergesort secuencialmente.

In [59]:
def mergesort_secuencial(v, n, num_trozos):

  tam_trozo = int(n / num_trozos)
  for i in range(0, num_trozos):
    izq = i*tam_trozo
    der = (i+1) * tam_trozo
    if i == num_trozos-1: der = n
    burbuja(v, izq, der)

	# Mezclar todos los trozos ordenados
  for i in range(num_trozos-2, -1, -1):
    izq=i*tam_trozo
    med=izq+tam_trozo
    mezclar(v,izq,med,n)

Probamos el algoritmo y medimos tiempos.

In [74]:
def ejercicio1(n=None, num_trozos=None):
  print("ORDENACIÓN MERGESORT SECUENCIAL\n")

  if not n or not num_trozos:
    n = int(input("Escriba el tamaño del vector: ")) # Número de datos que se quieren ordenar
    num_trozos = int(input("Escriba cúantos trozos quiere usar: ")) # Número de trozos que se quieren usar

  # Generación aleatoria de n datos
  datos = generar_datos(n)
  # Ordenar
  print("Ordenando...")
  ti=milisegundos() # tiempo inicial
  mergesort_secuencial(datos,n,num_trozos)
  tf=milisegundos() # tiempo final

  # Comprobar que está correcto
  res = esta_ordenado_array(datos, n, ti, tf)
  if res == -1:
    return tf-ti

In [67]:
ejercicio1()

ORDENACIÓN MERGESORT SECUENCIAL

Escriba el tamaño del vector: 10
Escriba cúantos trozos quiere usar: 2
Ordenando...
El array está ordenado


Tiempo invertido en ordenar: 0 milisegundos 




0

Teniendo esto ya, lo mejor es que calculemos los tiempos con los mismos datos que utilizábamos en la primera parte del trabajo. Recordemos que lo haciamos para tamaños de vectores 10, 1000, 10 000 y 1 000 000; y para trozos de tamaño 10, 100, 500 y 1000.

In [ ]:
n_tro = [10, 100, 500, 1000]
tam = [10, 100, 10000, 1000000]

resultado = []
for n in tam:
  fila = []
  for t in n_tro:
    if n == 1000000 and t in [10, 100, 500]:
      fila.append(">100 000")
    else:
      fila.append(ejercicio1(n, t))

  resultado.append(fila)

In [83]:
resultado = [ [abs(i) for i in j if isinstance(i, int)] for j in resultado]

In [84]:
resultado

[[0, 0, 2, 3], [0, 1, 12, 19], [894, 218, 724, 1354], [387758]]

Los resultado de un millón no aparecen porque los he saltado, ya que sobrepasaban demasiado el tiempo de cálculo. Pero tenemos la tabla.

Tabla | 10 | 100 | 500 | 1000
--- | --- | --- | --- | ---
10| 0 | 0 | 2 | 3
1000 | 0 | 1 | 12 | 19
10 000 | 894 | 218 | 724 | 1354
1 000 000 | >100 000 | >100 000 | >100 000 | 358 758

## Python paralelo

Para ello, vamos a utilizar la libreria multiprocessing y la clase pool().

In [11]:
def funcion_par(i, tam_trozo, num_trozos, n):
  izq = i*tam_trozo
  der = (i+1) * tam_trozo
  # Obtenemos el vector compartido para su consumición
  existing_shm = shared_memory.SharedMemory(name='mem_comp')
  v = np.ndarray((n,), dtype=np.int64, buffer=existing_shm.buf)
  if i == num_trozos-1: der = n
  burbuja(v, izq, der)

In [12]:
def mergesort_paralelo(v, n, num_trozos):

  pool = mp.Pool(mp.cpu_count())

  tam_trozo = int(n / num_trozos)
  [pool.apply(funcion_par, args=(i, tam_trozo, num_trozos, n)) for i in range(0, num_trozos)]

  pool.close()

	# Mezclar todos los trozos ordenados
  existing_shm = shared_memory.SharedMemory(name='mem_comp')
  aux = np.ndarray((n,), dtype=np.int64, buffer=existing_shm.buf)
  v[:] = aux[:]
  for i in range(num_trozos-2, -1, -1):
    izq=i*tam_trozo
    med=izq+tam_trozo
    mezclar(v,izq,med,n)

In [13]:
import multiprocessing as mp
from multiprocessing import shared_memory
import numpy as np

def ejercicio2(n=None, num_trozos=None):

  print("ORDENACIÓN MERGESORT PARALELO\n")

  if not n or not num_trozos:
    n = int(input("Escriba el tamaño del vector: ")) # Número de datos que se quieren ordenar
    num_trozos = int(input("Escriba cúantos trozos quiere usar: ")) # Número de trozos que se quieren usar

  # Generación aleatoria de n datos
  datos = generar_datos(n)
  # Compartimos el vector para que los procesos lo puedan consumir
  datos = np.array(datos)
  shm = shared_memory.SharedMemory(create=True, size=datos.nbytes, name="mem_comp")
  aux = np.ndarray(datos.shape, dtype=datos.dtype, buffer=shm.buf)
  aux[:] = datos[:]
  # Ordenar
  print("Ordenando...")
  ti=milisegundos() # tiempo inicial
  mergesort_paralelo(datos,n,num_trozos)
  tf=milisegundos() # tiempo final

  # Eliminamos la memoria compartida
  shm.close()
  shm.unlink()

  # Comprobar que está correcto
  res = esta_ordenado_array(datos, n, ti, tf)
  if res == -1:
    return tf-ti

In [168]:
ejercicio2()

ORDENACIÓN MERGESORT PARALELO

Escriba el tamaño del vector: 10
Escriba cúantos trozos quiere usar: 2
Ordenando...
El array está ordenado


Tiempo invertido en ordenar: 1215 milisegundos 




1215

Vemos que tarda bastante porque necesita compartir la memoria y generar los procesos.

In [14]:
n_tro = [10, 100, 500, 1000]
tam = [10, 100, 10000, 1000000]

resultado2 = []
for n in tam:
  fila = []
  for t in n_tro:
    if n == 1000000 and t in [10, 100, 500]:
      fila.append(">100 000")
    else:
      fila.append(ejercicio2(n, t))

  resultado2.append(fila)

ORDENACIÓN MERGESORT PARALELO

Ordenando...
El array está ordenado


Tiempo invertido en ordenar: 67 milisegundos 


ORDENACIÓN MERGESORT PARALELO

Ordenando...
El array está ordenado


Tiempo invertido en ordenar: 78 milisegundos 


ORDENACIÓN MERGESORT PARALELO

Ordenando...
El array está ordenado


Tiempo invertido en ordenar: 279 milisegundos 


ORDENACIÓN MERGESORT PARALELO

Ordenando...
El array está ordenado


Tiempo invertido en ordenar: 519 milisegundos 


ORDENACIÓN MERGESORT PARALELO

Ordenando...
El array está ordenado


Tiempo invertido en ordenar: 54 milisegundos 


ORDENACIÓN MERGESORT PARALELO

Ordenando...
El array está ordenado


Tiempo invertido en ordenar: 98 milisegundos 


ORDENACIÓN MERGESORT PARALELO

Ordenando...
El array está ordenado


Tiempo invertido en ordenar: 327 milisegundos 


ORDENACIÓN MERGESORT PARALELO

Ordenando...
El array está ordenado


Tiempo invertido en ordenar: 595 milisegundos 


ORDENACIÓN MERGESORT PARALELO

Ordenando...
El array está or

In [15]:
resultado2

[[67, 78, 279, 519],
 [54, 98, 327, 595],
 [5833, 645, 1928, 3502],
 ['>100 000', '>100 000', '>100 000', 578296]]

Tabla | 10 | 100 | 500 | 1000
--- | --- | --- | --- | ---
10| 67 | 78 | 279 | 519
1000 | 54 | 98 | 327 | 595
10 000 | 5833 | 645 | 1923 | 3502
1 000 000 | >100 000 | >100 000 | >100 000 | 578 296

## Numba

Para terminar esta parte vamos a realizar la implementación con Numba. Lo único que tenemos que cambiar por completo el código para poder usarlo desde Numba. Porque como no tenemos cachos u ordenamos todo con la burbuja o todo con merge. Voy a similar los cachos haciendo las divisiones pero tiene poco sentido.

In [11]:
from numba import cuda, njit, typed
import numpy as np

In [28]:
SMALL_MERGESORT_NUMBA = 40

@njit
def merge_numba(A, Aux, lo, mid, hi):

    i = lo
    j = mid + 1
    for k in range(lo, hi + 1):
        if i > mid:
            Aux[k] = A[j]
            j += 1
        elif j > hi:
            Aux[k] = A[i]
            i += 1
        elif A[j] < A[i]:
            Aux[k] = A[j]
            j += 1
        else:
            Aux[k] = A[i]
            i += 1


@njit
def insertion_sort_numba(A, lo, hi):

    for i in range(lo + 1, hi + 1):
        key = A[i]
        j = i - 1
        while (j >= lo) & (A[j] > key):
            A[j + 1] = A[j]
            j -= 1
        A[j + 1] = key


@njit
def merge_sort_numba(A, Aux, lo, hi):

    if hi - lo > SMALL_MERGESORT_NUMBA:
        mid = lo + ((hi - lo) >> 1)
        merge_sort_numba(Aux, A, lo, mid)
        merge_sort_numba(Aux, A, mid + 1, hi)
        if A[mid] > A[mid + 1]:
            merge_numba(A, Aux, lo, mid, hi)
        else:
            for i in range(lo, hi + 1):
                Aux[i] = A[i]
    else:
        insertion_sort_numba(Aux, lo, hi)

In [20]:
def ejercicio3(n=None):

  print("ORDENACIÓN MERGESORT NUMBA\n")

  if not n:
    n = int(input("Escriba el tamaño del vector: ")) # Número de datos que se quieren ordenar

  # Generación aleatoria de n datos
  datos = np.array(generar_datos(n))
  aux = np.copy(datos)
  # Ordenar
  print("Ordenando...")
  ti=milisegundos() # tiempo inicial
  merge_sort_numba(aux, datos, 0, len(datos) - 1)
  tf=milisegundos() # tiempo final

  # Comprobar que está correcto
  res = esta_ordenado_array(datos, n, ti, tf)
  if res == -1:
    return tf-ti

In [23]:
ejercicio3(10000)

ORDENACIÓN MERGESORT NUMBA

Ordenando...
El array está ordenado


Tiempo invertido en ordenar: 446 milisegundos 




446

Tenemos el código calculamos los tiempos para cada caso. Lo único que ahora no tiene ningún sentido hacer cachos evidentemente así que simplemente vamos a usar los tamaños.

In [24]:
tam = [10, 100, 10000, 1000000]

resultado3 = []
for n in tam:
  resultado3.append(ejercicio3(n))

ORDENACIÓN MERGESORT NUMBA

Ordenando...
El array está ordenado


Tiempo invertido en ordenar: 0 milisegundos 


ORDENACIÓN MERGESORT NUMBA

Ordenando...
El array está ordenado


Tiempo invertido en ordenar: 0 milisegundos 


ORDENACIÓN MERGESORT NUMBA

Ordenando...
El array está ordenado


Tiempo invertido en ordenar: 0 milisegundos 


ORDENACIÓN MERGESORT NUMBA

Ordenando...
El array está ordenado


Tiempo invertido en ordenar: 110 milisegundos 




In [25]:
resultado3

[0, 0, 0, 110]

Tabla | Resultados
--- | ---
10| 0
1000 | 0
10 000 | 0
1 000 000 | 110

Vemos que es de los mejores resultados que hemos obtenido hasta el momento. Pero solo utilizamos el algoritmo de mezcla, por eso vamos a hacer los cachos que comentaba al principio.

In [12]:
@njit
def burbuja_numba(v, izq, der):
  '''burbuja con numba'''
  aux = 0
  for i in range(0, der-izq-1):
    for j in range(izq, der-i-1):
      if v[j] > v[j+1]:
        aux = v[j]
        v[j]=v[j+1]
        v[j+1]=aux

@njit
def mezclar_numba(v , izq, med, der):
  '''Mezclar con numba'''

  aux= [0] * (der-izq)
  i=izq
  j=med
  k=0

  while i<med and j<der:
    if v[i] <= v[j]:
      aux[k] = v[i]
      i += 1
    else:
      aux[k] = v[j];
      j += 1
    k += 1

  while i<med:
    aux[k] = v[i]
    k += 1
    i += 1

  while j<der:
    aux[k] = v[j];
    k += 1
    j += 1

  for i in range(0, der-izq):
      v[i+izq] = aux[i]

@njit
def mergesort_numba(v, n, num_trozos):

  tam_trozo = int(n / num_trozos)
  for i in range(0, num_trozos):
    izq = i*tam_trozo
    der = (i+1) * tam_trozo
    if i == num_trozos-1: der = n
    burbuja_numba(v, izq, der)

	# Mezclar todos los trozos ordenados
  for i in range(num_trozos-2, -1, -1):
    izq=i*tam_trozo
    med=izq+tam_trozo
    mezclar_numba(v,izq,med,n)

In [13]:
def ejercicio3_1(n=None, num_trozos=None):
  print("ORDENACIÓN MERGESORT SECUENCIAL\n")

  if not n or not num_trozos:
    n = int(input("Escriba el tamaño del vector: ")) # Número de datos que se quieren ordenar
    num_trozos = int(input("Escriba cúantos trozos quiere usar: ")) # Número de trozos que se quieren usar

  # Generación aleatoria de n datos
  datos = typed.List(generar_datos(n))
  # Ordenar
  print("Ordenando...")
  ti=milisegundos() # tiempo inicial
  mergesort_numba(datos,n,num_trozos)
  tf=milisegundos() # tiempo final

  # Comprobar que está correcto
  res = esta_ordenado_array(datos, n, ti, tf)
  if res == -1:
    return tf-ti

In [14]:
ejercicio3_1(10, 5)

ORDENACIÓN MERGESORT SECUENCIAL

Ordenando...
El array está ordenado


Tiempo invertido en ordenar: 848 milisegundos 




848

Comprobamos tiempos.

In [15]:
n_tro = [10, 100, 500, 1000]
tam = [10, 100, 10000, 1000000]

resultado3_1 = []
for n in tam:
  fila = []
  for t in n_tro:
    if n == 1000000 and t in [10, 100, 500]:
      fila.append(">100 000")
    else:
      fila.append(ejercicio3_1(n, t))

  resultado3_1.append(fila)

ORDENACIÓN MERGESORT SECUENCIAL

Ordenando...
El array está ordenado


Tiempo invertido en ordenar: 0 milisegundos 


ORDENACIÓN MERGESORT SECUENCIAL

Ordenando...
El array está ordenado


Tiempo invertido en ordenar: 0 milisegundos 


ORDENACIÓN MERGESORT SECUENCIAL

Ordenando...
El array está ordenado


Tiempo invertido en ordenar: 1 milisegundos 


ORDENACIÓN MERGESORT SECUENCIAL

Ordenando...
El array está ordenado


Tiempo invertido en ordenar: 1 milisegundos 


ORDENACIÓN MERGESORT SECUENCIAL

Ordenando...
El array está ordenado


Tiempo invertido en ordenar: 0 milisegundos 


ORDENACIÓN MERGESORT SECUENCIAL

Ordenando...
El array está ordenado


Tiempo invertido en ordenar: 1 milisegundos 


ORDENACIÓN MERGESORT SECUENCIAL

Ordenando...
El array está ordenado


Tiempo invertido en ordenar: 2 milisegundos 


ORDENACIÓN MERGESORT SECUENCIAL

Ordenando...
El array está ordenado


Tiempo invertido en ordenar: 5 milisegundos 


ORDENACIÓN MERGESORT SECUENCIAL

Ordenando...
El array e

In [16]:
resultado3_1

[[0, 0, 1, 1],
 [0, 1, 2, 5],
 [372, 74, 193, 352],
 ['>100 000', '>100 000', '>100 000', 74181]]

Tabla | 10 | 100 | 500 | 1000
--- | --- | --- | --- | ---
10| 0 | 0 | 1 | 1
1000 | 0 | 1 | 2 | 5
10 000 | 372 | 74 | 193 | 352
1 000 000 | >100 000 | >100 000 | >100 000 | 74 181

Aún así notamos una gran mejoría.

Y con esto termina este apartado, ahora vamos a comparar estos resultados con los que ya teniamos de Java y C++.